st.c

```diff
@@ -2332,4 +2333,882 @@ rb_st_compact_table(st_table *tab)
+   during the search, return REBUILT_TABLE_ENTRY_IND.  */
+static st_index_t
+set_find_table_bin_ptr_and_reserve(set_table *tab, st_hash_t *hash_value,
+                               st_data_t key, st_index_t *bin_ind)
+{
+    int eq_p, rebuilt_p;
+    st_index_t ind;
+    st_hash_t curr_hash_value = *hash_value;
+#ifdef QUADRATIC_PROBE
+    st_index_t d;
+#else
+    st_index_t perturb;
+#endif
+    st_index_t entry_index;
+    st_index_t firset_deleted_bin_ind;
+    set_table_entry *entries;
+
+    ind = set_hash_bin(curr_hash_value, tab);
+#ifdef QUADRATIC_PROBE
+    d = 1;
+#else
+    perturb = curr_hash_value;
+#endif
+    firset_deleted_bin_ind = UNDEFINED_BIN_IND;
+    entries = tab->entries;
+    for (;;) {
+        entry_index = get_bin(tab->bins, set_get_size_ind(tab), ind);
+        if (EMPTY_BIN_P(entry_index)) {
+            tab->num_entries++;
+            entry_index = UNDEFINED_ENTRY_IND;
+            if (firset_deleted_bin_ind != UNDEFINED_BIN_IND) {
+                /* We can reuse bin of a deleted entry.  */
+                ind = firset_deleted_bin_ind;
+                MARK_SET_BIN_EMPTY(tab, ind);
+            }
+            break;
+        }
+        else if (! DELETED_BIN_P(entry_index)) {
+            DO_PTR_EQUAL_CHECK(tab, &entries[entry_index - ENTRY_BASE], curr_hash_value, key, eq_p, rebuilt_p);
+            if (EXPECT(rebuilt_p, 0))
+                return REBUILT_TABLE_ENTRY_IND;
+            if (eq_p)
+                break;
+        }
+        else if (firset_deleted_bin_ind == UNDEFINED_BIN_IND)
+            firset_deleted_bin_ind = ind;
+#ifdef QUADRATIC_PROBE
+        ind = set_hash_bin(ind + d, tab);
+        d++;
+#else
+        ind = set_secondary_hash(ind, tab, &perturb);
+#endif
+    }
+    *bin_ind = ind;
+    return entry_index;
+}
+
+/* Find an entry with KEY in table TAB.  Return non-zero if we found
+   it.  */
+int
+set_lookup(set_table *tab, st_data_t key)
+{
+    st_index_t bin;
+    st_hash_t hash = set_do_hash(key, tab);
+
+ retry:
+    if (tab->bins == NULL) {
+        bin = set_find_entry(tab, hash, key);
+        if (EXPECT(bin == REBUILT_TABLE_ENTRY_IND, 0))
+            goto retry;
+        if (bin == UNDEFINED_ENTRY_IND)
+            return 0;
+    }
+    else {
+        bin = set_find_table_entry_ind(tab, hash, key);
+        if (EXPECT(bin == REBUILT_TABLE_ENTRY_IND, 0))
+            goto retry;
+        if (bin == UNDEFINED_ENTRY_IND)
+            return 0;
+        bin -= ENTRY_BASE;
+    }
+    return 1;
+}
+
+/* Check the table and rebuild it if it is necessary.  */
+static inline void
+set_rebuild_table_if_necessary (set_table *tab)
+{
+    st_index_t bound = tab->entries_bound;
+
+    if (bound == set_get_allocated_entries(tab))
+        set_rebuild_table(tab);
+}
+
+/* Insert (KEY, VALUE) into table TAB and return zero.  If there is
+   already entry with KEY in the table, return nonzero and update
+   the value of the found entry.  */
```

Comment:
This comment need to be updated given there is no `VALUE`.

Reply:
Fixed this and the other cases.